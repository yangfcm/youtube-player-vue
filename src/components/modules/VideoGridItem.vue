<template>
  <div class="ui card app-video-card">
    <router-link :to="`/video/${videoId}`" class="image">
      <img
        :src="video.snippet.thumbnails.medium.url"
        :alt="video.snippet.title"
        :title="video.snippet.title"
      />
    </router-link>
    <div class="content">
      <div class="header app-video-title">
        <router-link
          :to="`/video/${videoId}`"
          :title="video.snippet.title"
          v-html="video.snippet.title"
        ></router-link>
      </div>
      <div class="meta">
        <router-link :to="`/channel/${video.snippet.channelId}`">{{ video.snippet.channelTitle }}</router-link>
      </div>
    </div>
    <div class="extra content">
      <span>
        <i class="calendar icon"></i>
        {{ video.snippet.publishedAt | date }}
      </span>
      <span v-if="video.statistics" class="right floated">
        <i class="eye icon"></i>
        {{ video.statistics.viewCount | decimal }}
      </span>
    </div>
  </div>
</template>

<script>
export default {
  props: ["video"],
  computed: {
    videoId() {
      return typeof this.video.id === "string"
        ? this.video.id
        : this.video.id.videoId;
    },
  },
};
</script>

<style scoped>
.app-video-card {
  width: 100%;
  height: 100%;
  margin: 0 !important;
}
.app-video-title {
  line-height: 23px;
  max-height: 46px;
  overflow: hidden;
  text-overflow: ellipsis;
}
</style>