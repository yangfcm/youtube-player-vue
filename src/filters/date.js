/**
 * Convert a date object to human-readable string
 */
export default (date) => {
  const dateObj = new Date(date);
  const day = dateObj.getDate();
  const month = new Intl.DateTimeFormat("en-US", {
    month: "short",
  }).format(dateObj);
  const year = dateObj.getFullYear();
  return `${day} ${month}, ${year}`;
};
