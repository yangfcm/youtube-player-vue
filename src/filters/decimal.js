/**
 * Separate an integer number by commas
 * e.g. 12345678 => 12,345,678
 */
export default (num) => {
  if (isNaN(Number(num))) return "";
  let strNum = String(num);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
