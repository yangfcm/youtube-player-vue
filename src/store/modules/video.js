import axios from "../../settings";
import { FETCH_VIDEOS, CATCH_VIDEO_ERROR, FETCH_VIDEO } from "../types";
import parseError from "../helpers/parseError";

const state = {
  videos: null,
  video: null,
  videoError: null,
};

const getters = {
  videoErrorMessage: (state) => {
    return parseError(state.videoError, "Failed to fetch video");
  },
};

const mutations = {
  FETCH_VIDEOS: (state, payload) => {
    state.videos = payload;
    state.videoError = null;
  },
  FETCH_VIDEO: (state, payload) => {
    state.video = payload;
    state.videoError = null;
  },
  CATCH_VIDEO_ERROR: (state, payload) => {
    state.videoError = payload;
  },
};

const actions = {
  /** Fetch a list of videos by filter */
  fetchVideos: async (context, filter, pageToken) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet,statistics",
          maxResults: 15,
          pageToken,
        },
      });
      context.commit(FETCH_VIDEOS, response.data);
    } catch (err) {
      context.commit(CATCH_VIDEO_ERROR, err);
    }
  },

  /** Fetch a video by id */
  fetchVideo: async (context, videoId) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: videoId,
        },
      });
      // console.log(response);
      context.commit(FETCH_VIDEO, response.data);
    } catch (err) {
      // console.log("error-==================", e);
      context.commit(CATCH_VIDEO_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
