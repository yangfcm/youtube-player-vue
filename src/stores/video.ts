import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { AsyncStatus } from '@/settings/types'
import { type VideoMetaSnippetStats, type VideoResponse } from './types'
import { fetchVideoAPI } from './api'
import { type AxiosResponse } from 'axios'

type VideoStore = {
  error: string
  status: AsyncStatus
  video: Record<string, VideoMetaSnippetStats>
}

export const useVideoStore = defineStore('video', () => {
  const videoStore = ref<VideoStore>({
    error: '',
    status: AsyncStatus.IDLE,
    video: {},
  })

  const status = computed(() => videoStore.value.status)
  const error = computed(() => videoStore.value.error)
  const video = computed(() => videoStore.value.video || {})

  const fetchVideo = async (videoId: string) => {
    try {
      videoStore.value.status = AsyncStatus.LOADING
      videoStore.value.error = ''
      const response: AxiosResponse<VideoResponse> = await fetchVideoAPI(videoId)
      const video = response.data.items[0]
      if (video) {
        videoStore.value.video = {
          ...(videoStore.value.video || {}),
          [video.id as string]: video,
        }
      }
      videoStore.value.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      videoStore.value.status = AsyncStatus.FAIL
      videoStore.value.error = err.message
    }
  }

  return {
    videoState: videoStore,
    video,
    status,
    error,
    fetchVideo,
  }
})
