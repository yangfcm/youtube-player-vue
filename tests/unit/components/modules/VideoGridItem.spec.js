import { shallowMount, RouterLinkStub, createLocalVue } from "@vue/test-utils";
import VideoGridItem from "@/components/modules/VideoGridItem";
import { videoItem1 as video } from "../../store/fixtures/video";
import date from "@/filters/date";
import decimal from "@/filters/decimal";

describe("Test VideoGridItem component", () => {
  const localVue = createLocalVue();
  localVue.filter("date", date);
  localVue.filter("decimal", decimal);
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(VideoGridItem, {
      propsData: {
        video,
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
      localVue,
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render correct content", () => {
    expect(wrapper.find("img").attributes("src")).toBe(
      video.snippet.thumbnails.medium.url
    );
    expect(wrapper.find("a.image").props().to).toBe(`/video/${video.id}`);
    expect(wrapper.find("div.app-video-title a").props().to).toBe(
      `/video/${video.id}`
    );
    expect(wrapper.find("div.app-video-title").text()).toContain(
      video.snippet.title
    );
    expect(wrapper.find("div.meta a").props().to).toBe(
      `/channel/${video.snippet.channelId}`
    );
  });
});
