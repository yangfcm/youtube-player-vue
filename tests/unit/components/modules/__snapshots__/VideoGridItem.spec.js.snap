// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoGridItem component should match snapshot 1`] = `
<div class="ui card app-video-card"><a class="image"><img src="https://test-thumbnail-medium-url.mock" alt="Mock video title item1" title="Mock video title item1"></a>
  <div class="content">
    <div class="header app-video-title"><a title="Mock video title item1">Mock video title item1</a></div>
    <div class="meta"><a>Ford Motor Company</a></div>
  </div>
  <div class="extra content"><span><i class="calendar icon"></i>
      14 Jul, 2020
    </span> <span class="right floated"><i class="eye icon"></i>
      1,766,115
    </span></div>
</div>
`;
